Use E0665 for missing `#[default]` on enum and update doc

estebank opened the pull request:
The docs for E0665 when doing `#[derive(Default]` on an `enum` previously didn't mention `#[default]` at all, or made a distinction between unit variants, that can be annotated, and tuple or struct variants, which cannot.

E0665 was not being emitted, we now use it for the same error it belonged to before.

```
error[E0665]: `#[derive(Default)]` on enum with no `#[default]`
  --> $DIR/macros-nonfatal-errors.rs:42:10
   |
LL |   #[derive(Default)]
   |            ^^^^^^^
LL | / enum NoDeclaredDefault {
LL | |     Foo,
LL | |     Bar,
LL | | }
   | |_- this enum needs a unit variant marked with `#[default]`
   |
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)
help: make this unit variant default by placing `#[default]` on it
   |
LL |     #[default] Foo,
   |     ++++++++++
help: make this unit variant default by placing `#[default]` on it
   |
LL |     #[default] Bar,
   |     ++++++++++
```

estebank committed:
Mention `#[default]` in E0655 code index

fmease commented on compiler/rustc_error_codes/src/error_codes/E0665.md:
This error code is no longer emitted by the compiler and I don't see the point in updating this entry if it never sees the light of day.

The current diagnostic emitted on an `#[derive(Default)] enum` without `#[default]` doesn't have any error code associated with it. I don't know why it didn't take over E0665. If you make it use E0665 (I think that makes sense), then I'm fine with updating this entry.

fmease commented on compiler/rustc_error_codes/src/error_codes/E0665.md:
```diff
@@ -16,18 +16,30 @@ The `Default` cannot be derived on an enum for the simple reason that the
 
-If you still want to implement `Default` on your enum, you'll have to do it "by
-hand":
+For the case where the desired default variant has no data, you can annotate
```
```suggestion
For the case where the desired default variant has no payload, you can annotate
```

Minor, semantics: "Data" is pretty vague and ambiguous. For example, you are allowed to place `#[default]` on a (unit) variant with an explicit discriminant like `Var = 16`. Arguably, `Var` does have *data*. I'd prefer we use the term *payload* here or rephrase it slightly to mention *unit variant* instead.

fmease commented on compiler/rustc_error_codes/src/error_codes/E0665.md:
````diff
@@ -16,18 +16,30 @@ The `Default` cannot be derived on an enum for the simple reason that the
+```
+
+In the case where the default variant does have data, you will have to
+implement `Default` on your enum "by hand":
````
```suggestion
implement `Default` on your enum manually:
```

?

fmease commented on compiler/rustc_error_codes/src/error_codes/E0665.md:
````diff
@@ -16,18 +16,30 @@ The `Default` cannot be derived on an enum for the simple reason that the
 }
+```
+
+In the case where the default variant does have data, you will have to
````
```suggestion
In the case where the default variant does have payload, you will have to
```

ditto

estebank committed:
Use E0665 for missing `#[default]` error

Use orphaned error code for the same error it belonged to before.

```
error[E0665]: `#[derive(Default)]` on enum with no `#[default]`
  --> $DIR/macros-nonfatal-errors.rs:42:10
   |
LL |   #[derive(Default)]
   |            ^^^^^^^
LL | / enum NoDeclaredDefault {
LL | |     Foo,
LL | |     Bar,
LL | | }
   | |_- this enum needs a unit variant marked with `#[default]`
   |
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)
help: make this unit variant default by placing `#[default]` on it
   |
LL |     #[default] Foo,
   |     ~~~~~~~~~~~~~~
help: make this unit variant default by placing `#[default]` on it
   |
LL |     #[default] Bar,
   |     ~~~~~~~~~~~~~~
```

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12441488461/job/34738441791) [(plain)](https://github.com/rust-lang/rust/commit/bac7682922ed03153b605ce8ee81a714259f86ac/checks/34738441791/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.872 Building wheels for collected packages: reuse
#13 2.873   Building wheel for reuse (pyproject.toml): started
#13 3.121   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.122   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.122   Stored in directory: /tmp/pip-ephem-wheel-cache-rxvzwn7e/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.125 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.526 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.527 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.057 Collecting virtualenv
#13 4.057 Collecting virtualenv
#13 4.125   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.328      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 21.4 MB/s eta 0:00:00
#13 4.386 Collecting platformdirs<5,>=3.9.1
#13 4.393   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.434 Collecting filelock<4,>=3.12.2
#13 4.463 Collecting distlib<1,>=0.3.7
#13 4.470   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.484      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 39.0 MB/s eta 0:00:00
#13 4.484      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 39.0 MB/s eta 0:00:00
#13 4.565 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.751 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      126912 kB
DirectMap2M:     7213056 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
##[endgroup]
fmt check
fmt: checked 5736 files
tidy check
tidy error: `/checkout/compiler/rustc_error_codes/src/error_codes/E0665.md` doesn't use its own error code in compile_fail example
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sat Dec 21 02:52:28 UTC 2024
  network time: Sat, 21 Dec 2024 02:52:28 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

fmease commented on compiler/rustc_builtin_macros/src/errors.rs:
```diff
@@ -369,21 +369,21 @@ pub(crate) struct DerivePathArgsValue {
+#[derive(Subdiagnostic, Debug)]
 #[suggestion(
     builtin_macros_suggestion,
     code = "#[default] {ident}",
```
Preexisting: Ideally, this suggestion would be `(v.span.shrink_to_hi(), "#[default] ")` instead of `(v.span, format!("#[defaukt] {ident}"))` like it is right now. This would also allow us to get rid of the `pub(crate) ident: Ident,`.

If you want to make the change feel free to do so in this PR :)

fmease reviewed (approved):
Thanks, this diagnostic and the error index entry is now so much better!

r=me after nits & CI failure addressed and title change

estebank commented on compiler/rustc_error_codes/src/error_codes/E0665.md:
````diff
@@ -1,10 +1,9 @@
 Erroneous code example:
 
-```compile_fail
+```compile_fail,ignore
````
I had to do this to get over the stage0 test of this page.

fmease replied:
Doesn't the following make this work?

````md
```compile_fail,E0665
````

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12441763542/job/34739094603) [(plain)](https://github.com/rust-lang/rust/commit/4bf894890752cdea3bb52d7b9ed364b225d36e28/checks/34739094603/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 28.1s done
#21 DONE 42.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
diff of stderr:

61 help: make this unit variant default by placing `#[default]` on it
62    |
63 LL |     #[default] Foo,
+    |     ++++++++++
65 help: make this unit variant default by placing `#[default]` on it
66    |
66    |
67 LL |     #[default] Bar,
-    |     ~~~~~~~~~~~~~~
+    |     ++++++++++
69 
70 error[E0665]: `#[derive(Default)]` on enum with no `#[default]`
---
To only update this specific test, also pass `--test-args macros/macros-nonfatal-errors.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/macros-nonfatal-errors.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/macros-nonfatal-errors" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
--- stderr -------------------------------
error: the `#[default]` attribute may only be used on unit enum variants
   |
   |
LL |     #[default] //~ ERROR the `#[default]` attribute may only be used on unit enum variants
   |
   = help: consider a manual implementation of `Default`


error: the `#[default]` attribute may only be used on unit enum variants
   |
   |
LL | struct DefaultInnerAttrTupleStruct(#[default] ());
   |
   = help: consider a manual implementation of `Default`


error: the `#[default]` attribute may only be used on unit enum variants
   |
   |
LL | #[default] //~ ERROR the `#[default]` attribute may only be used on unit enum variants
   |
   = help: consider a manual implementation of `Default`


error: the `#[default]` attribute may only be used on unit enum variants
   |
   |
LL | #[default] //~ ERROR the `#[default]` attribute may only be used on unit enum variants
   |
   = help: consider a manual implementation of `Default`


error: the `#[default]` attribute may only be used on unit enum variants
   |
   |
LL |     Foo = #[default] 0, //~ ERROR the `#[default]` attribute may only be used on unit enum variants
   |
   = help: consider a manual implementation of `Default`


error: the `#[default]` attribute may only be used on unit enum variants
   |
   |
LL |     Bar([u8; #[default] 1]), //~ ERROR the `#[default]` attribute may only be used on unit enum variants
   |
   = help: consider a manual implementation of `Default`

error[E0665]: `#[derive(Default)]` on enum with no `#[default]`
error[E0665]: `#[derive(Default)]` on enum with no `#[default]`
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:42:10
   |
LL |   #[derive(Default)] //~ ERROR `#[derive(Default)]` on enum with no `#[default]`
LL | / enum NoDeclaredDefault {
LL | |     Foo,
LL | |     Bar,
LL | | }
LL | | }
   | |_- this enum needs a unit variant marked with `#[default]`
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)
help: make this unit variant default by placing `#[default]` on it
   |
   |
LL |     #[default] Foo,
help: make this unit variant default by placing `#[default]` on it
   |
   |
LL |     #[default] Bar,

error[E0665]: `#[derive(Default)]` on enum with no `#[default]`
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:48:10
   |
   |
LL |   #[derive(Default)] //~ ERROR `#[derive(Default)]` on enum with no `#[default]`
   |            ^^^^^^^
LL | / enum NoDeclaredDefaultWithoutUnitVariant {
LL | |     Foo(i32),
LL | |     Bar(i32),
LL | | }
   | |_- this enum needs a unit variant marked with `#[default]`
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)

error: multiple declared defaults
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:54:10
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:54:10
   |
LL | #[derive(Default)] //~ ERROR multiple declared defaults
...
LL |     Foo,
   |     --- first default
LL |     #[default]
---
   |
   = note: only one variant can be default
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)

error: `#[default]` attribute does not accept a value
   |
   |
LL |     #[default = 1] //~ ERROR `#[default]` attribute does not accept a value
   |
   = help: try using `#[default]`

error: multiple `#[default]` attributes
error: multiple `#[default]` attributes
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:74:5
   |
LL |     #[default]
   |     ---------- `#[default]` used here
LL |     #[default]
   |     ---------- `#[default]` used again here
LL |     Foo, //~ERROR multiple `#[default]` attributes
   |
   |
   = note: only one `#[default]` attribute is needed
  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:73:5
   |
LL |     #[default]
   |     ^^^^^^^^^^
   |     ^^^^^^^^^^

error: multiple `#[default]` attributes
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:84:5
   |
LL |     #[default]
   |     ---------- `#[default]` used here
LL |     #[default]
   |     ---------- `#[default]` used again here
...
LL |     Foo, //~ERROR multiple `#[default]` attributes
   |
   |
   = note: only one `#[default]` attribute is needed
  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:81:5
   |
LL |     #[default]
   |     ^^^^^^^^^^
   |     ^^^^^^^^^^
LL |     #[default]
   |     ^^^^^^^^^^
LL |     #[default]
   |     ^^^^^^^^^^

error: the `#[default]` attribute may only be used on unit enum variants
   |
   |
LL |     Foo {}, //~ ERROR the `#[default]` attribute may only be used on unit enum variants
   |
   = help: consider a manual implementation of `Default`

error: default variant must be exhaustive
error: default variant must be exhaustive
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:99:5
   |
LL |     #[non_exhaustive]
   |     ----------------- declared `#[non_exhaustive]` here
LL |     Foo, //~ ERROR default variant must be exhaustive
   |
   = help: consider a manual implementation of `Default`

error: asm template must be a string literal
error: asm template must be a string literal
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:104:10
   |
LL |     asm!(invalid); //~ ERROR


error: `concat_idents!()` requires ident args
   |
   |
LL |     concat_idents!("not", "idents"); //~ ERROR

error: argument must be a string literal
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:109:17
   |
   |
LL |     option_env!(invalid); //~ ERROR

error: expected string literal
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:110:10
   |
   |
LL |     env!(invalid); //~ ERROR


error: `env!()` takes 1 or 2 arguments
   |
   |
LL |     env!(foo, abr, baz); //~ ERROR

error: environment variable `RUST_HOPEFULLY_THIS_DOESNT_EXIST` not defined at compile time
##[error]  --> /checkout/tests/ui/macros/macros-nonfatal-errors.rs:112:5
   |
   |
LL |     env!("RUST_HOPEFULLY_THIS_DOESNT_EXIST"); //~ ERROR
   |
```

</details>

estebank commented:
@bors r=fmease

bors commented:
:pushpin: Commit 94812f1c8f455579a3a1c127c2bbac48fc7cecef has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 94812f1c8f455579a3a1c127c2bbac48fc7cecef -->
<!-- homu: {"type":"Approved","sha":"94812f1c8f455579a3a1c127c2bbac48fc7cecef","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
